Reformatted instructions for the first four ES6 challenges

BrendanSweeny opened the pull request:
<!-- freeCodeCamp Pull Request Template -->

<!-- IMPORTANT Please review https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/CONTRIBUTING.md for detailed contributing guidelines -->
<!-- Help with PRs can be found at https://gitter.im/FreeCodeCamp/Contributors -->
<!-- Make sure that your PR is not a duplicate -->

#### Pre-Submission Checklist
<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply. -->
<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change
<!-- What type of change does your code introduce? After creating the PR, tick the checkboxes that apply. -->
- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:
<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask in the Help Contributors room linked above. We're here to help! -->
- [x] Tested changes locally.
- [ ] Closes currently open issue (replace XXXX with an issue no): Closes #XXXX

Progress towards issue #12698, which is a compiled list of QA tasks for the new ES6 section introduced in beta

#### Description
<!-- Describe your changes in detail -->
* Updated instructions to fixed spelling/grammar mistakes and performed minor rewording for clarification.
* Improved overall formatting of instructions by adding multi-line blockquotes for example code and by formatting inline code to identify variables and keywords in the instructions text.
* Changed titles of challenges to better conform to the format used in the titles in the Basic Javascript section where appropriate.
* Where appropriate, made minor changes to the challengeSeed syntax to conform with freeCodeCamp style guidelines (e.g. added a few semicolons, added spaces between function statement and function block, etc.) without changing functionality.
* Made no improvements to the functionality of the tests for any of the challenges. That is still needed in some cases.

BrendanSweeny committed:
Reformat first four ES6 challenge instructions

Greenheart commented on seed/challenges/02-javascript-algorithms-and-data-structures/es6.json:
```diff
@@ -71,66 +67,29 @@
+        "<strong>Note</strong><br>Remember to add \"use strict\" to the top of your code."
       ],
       "challengeSeed": [
         "var newCampers = [{camper: \"Wil\"}, {camper: \"Sam\"},{camper: \"Dav\"}];",
```
Sorry if I'm being too picky. There's a missing space between Sam and Dav

Greenheart commented on seed/challenges/02-javascript-algorithms-and-data-structures/es6.json:
```diff
@@ -189,29 +144,25 @@
+        "To make an object immutable, you can use <code>Object.freeze()</code>.",
+        "<hr>",
+        "An array is delcared as <code>const s = [5, 7, 2]</code>. Change the array to <code>[2, 5, 7]</code>.",
+        "<strong>Note</strong><br>Don't forget to use strict mode."
```
Possibly change format to `<code>"use strict";</code>` here?

Greenheart commented on seed/challenges/02-javascript-algorithms-and-data-structures/es6.json:
```diff
@@ -71,66 +67,29 @@
+        "for (var i = 0; i < newCampers.length; i++) {",
         "// only change code above this line",
-        "    newCampers[i].roleCall = function(){",
+        "    newCampers[i].roleCall = function() {",
```
This might be outside the scope of your change, but I wanted to at least add my notes here.

The editor shows an error for this line, stating that we shouldn't create functions within loops. We can avoid this by moving this function definition out of the loop and only reference it in the loop.

Updated `challengeSeed`:
```js
var newCampers = [{camper: "Wil"}, {camper: "Sam"}, {camper: "Dav"}];
function roleCall (i) {
  return "Camper # " + (i + 1) + " has arrived.";
} 

// only change code below this line
for (var i = 0; i < newCampers.length; i++) {
// only change code above this line
  newCampers[i].roleCall = roleCall.bind(null, i);
}

console.log(newCampers[0].roleCall());
console.log(newCampers[1].roleCall());
console.log(newCampers[2].roleCall());
```

However, this also means we lose the closure that gives us access to `i`. Since code isn't supposed to be written like this anyway, I think we're better of by using `roleCall.bind(null, i)`, despite the increased complexity. The code above works just like the previous example, but uses better (if not best) practices.

Greenheart replied:
@erictleung What do you think about this?

BrendanSweeny replied:
I tend to agree that we should avoid using an example that also represents poor programming practice. How about adding another `console.log(typeof i);` to your proposed challengeSeed and changing the instructions to something like, "Fix the code so each camper shows their correct index when they are called, and without declaring i as a global variable."? It's a simple fix. It alerts the user to the different scopes by logging undefined or "number", and the solution is still the same. For now, I am going to hold off updating it until there is a final decision.

Greenheart replied:
@BrendanSweeny Where should we add that `console.log(typeof i);` in that case? Inside the the `roleCall` function? To the end of the `challengeSeed`? :blush:

@freeCodeCamp/moderators What do you think about this example?

BrendanSweeny replied:
My idea would be to add it to the end like so:
```
var newCampers = [{camper: "Wil"}, {camper: "Sam"}, {camper: "Dav"}];
function roleCall (i) {
  return "Camper # " + (i + 1) + " has arrived.";
} 

// only change code below this line
for (var i = 0; i < newCampers.length; i++) {
// only change code above this line
  newCampers[i].roleCall = roleCall.bind(null, i);
}

console.log(newCampers[0].roleCall());
console.log(newCampers[1].roleCall());
console.log(newCampers[2].roleCall());
console.log(typeof i);
```
This would alert that user that `i` was defined in the global scope or not by either logging "number" with `var `or "undefined" `let`. The more important part, that I neglected to mention in my previous post, was to add a test to check for this. My first attempts to come up with a working test using `assert(typeof i === "undefined", "message...");` or `assert.isUndefined(i, "message...");` have not succeeded, so I don't have a solution at this point.

Greenheart replied:
@BrendanSweeny Ah, that's a great idea! The test case would make it even clearer.

As @raisedadead pointed out below, let's continue this discussion in a separate issue and merge your changes so far. Thanks! :smile:

Greenheart reviewed (changes requested):
@BrendanSweeny Nice work on these fixes!

Some general thoughts:
* Indentation for code examples: Should it be 4 or 2? I think 2 is common in other examples. Could you fix this while you're at it?

erictleung commented on seed/challenges/02-javascript-algorithms-and-data-structures/es6.json:
```diff
@@ -36,25 +36,21 @@
+        "Let's try using the <code>let</code> keyword.",
+        "<hr>",
+        "Fix the code so that it only uses the <code>let</code> keyword and makes the errors go away.",
+        "<strong>Note</strong><br>Remember to add \"use strict\" to the top of your code."
```
You'll probably still want those escape characters `<code>\"use strict\";</code>`

Greenheart replied:
Nice catch!

BrendanSweeny commented:
Good eye, @Greenheart ! Thanks for the feedback. I'll go ahead and format the various "use strict;" instances, correct the "Camper" indices and missing space. You are right about the two spaces vs four per the freeCodeCamp style guide, so I'll check each block of example code more closely.

BrendanSweeny commented:
I have amended my commit with the following changes:

- Added <code>\"use strict\"</code> where necessary.
- Fixed the indices for the newCampers array in the test messages and the missing space between Sam and Dav.
- Changed 4-space indents to 2-space indents in blockquotes in the challenge instructions and the challengeSeeds where necessary.

Thanks for taking a look!

Greenheart reviewed (commented):
@BrendanSweeny Thanks for the update! This is looking great but I'm waiting to merge this for now as I think we should figure out what we want to do with the `roleCall` example in "Compare Scopes of the var and let keywords".

Hopefully we'll at the very least be able to merge this in the next few days :smile:

raisedadead commented:
@Greenheart just a thought, can we not merge this and track adding/modifying the example code in a separate issue?

QuincyLarson commented:
@BrendanSweeny thanks for your help with these!
